### PR TITLE
feat(driver): rewrite js2native.js with c++

### PR DESCRIPTION
**Pull request by ilikethese**
# Pre-PR Checklist

- [x] I added/updated relevant documentation.
- [x] I followed the [Convention Commit](https://conventionalcommits.org/) guideline with maximum 72 characters to submit commit message.
- [x] I squashed the repeated code commits.
- [x] I signed the [CLA].
- [x] I added/updated test cases to check the change I am making.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: driver/js/src/napi/jsc/js_native_api_jsc.cc [hunk 1]
```diff
@@ -515,6 +515,60 @@ void JSCCtx::RegisterNativeBinding(const string_view& name,
...
+  void *private_data = static_cast<void *>(tuple.get());
+  function_private_data_container_.push_back(std::move(tuple));
+  JSClassDefinition cls_def = kJSClassDefinitionEmpty;
+  cls_def.callAsFunction = native_func_callback;
+  JSClassRef cls_ref = JSClassCreate(&cls_def);
+  JSObjectRef func_object = JSObjectMake(context_, cls_ref, private_data);
+  JSClassRelease(cls_ref);
+  bool ret = JSValueIsObject(context_, value_ref);
+  if(ret) {
+    JSObjectRef object_ref = JSValueToObject(context_, value_ref, NULL);
```
Comment (churchill-zhang) [inline driver/js/src/napi/jsc/js_native_api_jsc.cc:566]: native_func_name 如果只用在这里，初始化也放这里把

File: driver/js/src/napi/jsc/js_native_api_value_jsc.cc [hunk 2]
```diff
@@ -372,6 +374,163 @@ std::shared_ptr<CtxValue> JSCCtx::CreateByteBuffer(const void* buffer, size_t le
...
+}
+
+std::shared_ptr<CtxValue> JSCCtx::CreateRejectPromise(
+  const std::shared_ptr<CtxValue>& value)  {
+  JSObjectRef global = JSContextGetGlobalObject(context_);
+  JSStringRef name = JSStringCreateWithUTF8CString("Promise");
+  JSValueRef exception = nullptr;
+  JSValueRef value_ref = JSObjectGetProperty(context_, global, name, &exception);
+  if (exception) {
+    SetException(std::make_shared<hippy::napi::JSCCtxValue>(context_, exception));
```
Comment (churchill-zhang) [inline driver/js/src/napi/jsc/js_native_api_value_jsc.cc:443]: name 泄露了

File: driver/js/src/napi/jsc/js_native_api_value_jsc.cc [hunk 3]
```diff
@@ -372,6 +374,163 @@ std::shared_ptr<CtxValue> JSCCtx::CreateByteBuffer(const void* buffer, size_t le
...
+}
+
+std::shared_ptr<CtxValue> JSCCtx::CreatePromiseWithCallback(
+  PromiseCallback callback) {
+  JSObjectRef global = JSContextGetGlobalObject(context_);
+  JSStringRef name = JSStringCreateWithUTF8CString("Promise");
+  JSValueRef exception = nullptr;
+  JSValueRef value_ref = JSObjectGetProperty(context_, global, name, &exception);
+  if (exception) {
+    SetException(std::make_shared<hippy::napi::JSCCtxValue>(context_, exception));
```
Comment (churchill-zhang) [inline driver/js/src/napi/jsc/js_native_api_value_jsc.cc:498]: name 泄露同上，下不说明一起改了吧。最好哪里使用哪里Create，不然很容易遗漏

File: framework/android/src/main/cpp/src/bridge/js2native.cc [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    auto error = ctx->CreateTypeError("hippyCallNatives not defined");
+    return ctx->CreateRejectPromise(error);
+  }
+
+  if (data_tuple.count_ < 2) {
+    auto error = ctx->CreateTypeError("callNativeWithPromise arguments length must be larger than 2");
+    return ctx->CreateRejectPromise(error);
+  }
+
+  PromiseCallback callback = [ctx, data_tuple](const std::shared_ptr<CtxValue>& resolve,
```
Comment (churchill-zhang) [inline framework/android/src/main/cpp/src/bridge/js2native.cc:163]: 理论promise是异步的，这里ctx不是weak会有点问题吧

File: framework/android/src/main/cpp/src/bridge/js2native.cc [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+  auto global_proto = ctx->GetGlobalObjVar(kGlobalProto);
+  auto module_call_id = ctx->GetProperty(global_proto, kModuleCallId);
+  int32_t current_call_id;
+  ctx->GetValueNumber(module_call_id, &current_call_id);
+  ctx->SetProperty(global_proto, kModuleCallId, ctx->CreateNumber(current_call_id + 1), PropertyAttribute::None);
+
+  bool has_callback_function = false;
+  std::vector<std::shared_ptr<CtxValue>> param_vector;
+
+  for (size_t i = 3; i < data_tuple.count_; i++) {
```
Comment (churchill-zhang) [inline framework/android/src/main/cpp/src/bridge/js2native.cc:262]: 估计要写下注释，说下1、2、3到底代表什么，光看代码是一定看不懂的。。。

File: framework/android/src/main/cpp/src/bridge/js2native.cc [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    return false;
+  }
+
+  auto bridge = ctx->GetProperty(hippy, kBridge);
+  if (ctx->IsNullOrUndefined(bridge)) {
+    FOOTSTONE_DLOG(INFO) << "Js To Native Bridge Object is null or undefined.";
+    return false;
+  }
+
+  RegisterFunction call_native = [ctx](void* data) {
```
Comment (churchill-zhang) [inline framework/android/src/main/cpp/src/bridge/js2native.cc:350]: lambda捕获都要注意下生命周期问题

File: driver/js/lib/bridge/android/js2native.js [hunk 7]
```diff
@@ -27,145 +27,4 @@
  *  1 : naive call callback, delete cache from moduleCallList by js
```
Comment (zoomchan-cxj) [inline driver/js/lib/bridge/android/js2native.js:27]: 不应该再保留js文件，最好建立一个js和C++的规则标准

**Pull request closed without merging**